Gun support. (ABS_* support)

nadenislamarre opened the pull request:
absolute mouses must be detected as pointer so that gun can be configured and work.

the patch comes from the batocera project
https://github.com/batocera-linux/batocera.linux/blob/master/package/batocera/emulators/dolphin-emu/008-guns.patch

and a demonstration of it can be seen here :
https://www.youtube.com/watch?v=_EvzlRD4Ttg

Signed-off-by: Nicolas Adenis-Lamarre <nicolas.adenis.lamarre@gmail.com>

JosJuice commented:
@dolphin-emu-bot rebuild

Pokechu22 commented:
Hmm, does `INPUT_PROP_BUTTONPAD` even make sense here? [The docs](https://docs.kernel.org/next/input/event-codes.html#input-prop-buttonpad) say it's for touchpads that are also a button, while `INPUT_PROP_POINTER` is [for touchpads and mice](https://docs.kernel.org/next/input/event-codes.html#input-prop-direct-input-prop-pointer). Did using a regular mouse as an input not work before?

(cc @jordan-woyak who added this in #8473)

nadenislamarre commented:
yes, for me, INPUT_PROP_BUTTONPAD is strange.
i've not tested a regular mouse, but from comments have seen, it should work.
but a gun is not a mouse. it has absolute coordonnates, not relative one.
the code to reach is this one : if (libevdev_has_event_code(dev, EV_ABS, code))

however, not that my patch is maybe not complete. We are doing some new test, and while it works for 4/3 games, it seems to cause issues on 16/9 games (or the contrary, i'm no sure of what was reported). This is generally an issue with some emulators that i've to fix, on the x axis : 0 on the game is not 0 on the gun axis (because of the blacks borders).

note that while i've read the code for various emulators to make guns working, i've not gone deeply in each one. it includes dolphin-emu. So, take attention before merging. however, for this simple test, it cannot break anything, just let more device to work.
As a general rule, personally, i would never do the tests you did : if the device has a abs coordonates, let it configure it ; don't add extra "if" that may disable the axis of a device. There are very strange devices (like the wiimote for the most known) that may not work because of the "if" while they have abs axis.

nadenislamarre commented:
forget that patch for the moment. i'm reworking on it. i found several points i don't like in the dolphin linux event code, i will try to make it better.

nadenislamarre commented:
i've finally changed my analysis.
The issues comes from the codes inside : if (is_pointing_device) { * }
making some device not correctly analysed mainly because of the "return true".
The code assumes things (because of the return) about devices (like if other case must not be parsed).

Moreover, the code "for (auto remove_prefix : {"BTN_", "KEY_"})" is problematic.
mainly for some cases like my virtual pad having both BTN_LEFT and KEY_LEFT.
same, this code assume that a device cannot have both, just for an esthetic thing.
I think it should be removed too.

nadenislamarre commented:
with all that. now guns are working on batocera.
finally, it was removing hacks that solved it (i coded no line)

Pokechu22 commented:
@dolphin-emu-bot rebuild

jordan-woyak commented:
The hacks were in place to prevent input name clashes when multiple evdev nodes are combined. Also see the comment above `has_sensible_button_names` (which is now unused). The "names" are often useless for gamepad devices.

I suspect you fixed your gun and broke other devices.

Is your issue ultimately that you have both a `BTN_LEFT` and a `KEY_LEFT`?
Are they on the same "node" or multiple evdev nodes that Dolphin combines?

nadenislamarre commented:
hum, i didn't know that events could be combined in dolphin. how do you do that ? In my case, i use only one event.
however, i don't get the link with the BTN_ KEY_ removals which looks more esthetic to me (but maybe i'm wrong).
"The "names" are often useless for gamepad devices." : maybe in dolphin, but on several programs it is usefull. and as a general rule, i prefer named that indexed number that are more or less random and not consistant from one library to an other contrary to codes.

"I suspect you fixed your gun and broke other devices."
i fixed my events yes, but i removed some bad hacks (my opinion, i let you decide) too in the code.
i've worked for 6 years on events on about 30 emulators / games / libraries (sdl and so on). i've added wii guitar hero instruments in the linux kernel. evdev gun support to mame, i've worked on pads/guns support in the batocera distribution. i think i've a big experience about events. I didn't just fix my device. i tried to add my experience to remove some hacks (that break my deivices to work). but my dolphin experience is small. so yes, maybe i'm wrong in my patch ; i let you decide, but i'm quite sure, the lines are remove are not good.

jordan-woyak commented:
e.g. evdev separates buttons from motion input from touchpad on the PlayStation 4 controller. Dolphin attempts to combine these for more convenient configuration. The combining isn't configurable. It's all in code based on having a common parent device or some such thing. I have some local changes that attempt to improve things by appropriately prefixing input names with things like "Touchpad " and maybe this would improve your gun situation.

The `BTN_` `KEY_` prefix removals are purely aesthetic, yes. Eliminating that code will break existing configs, but that's fine if we determine that it actually needs to be done.

When I say the names were useless, I don't mean "Dolphin doesn't need names". I mean they were blatantly wrong and confusing. I don't have a specific example at the moment, but something like a "B" button on a gamepad would acquire the name "DELETE", along with something equivalently stupid for each button, without any pattern.

I want to fix your gun without breaking other devices but I'm still not sure what your actual issue is/was. Were inputs not being populated for your gun?

nadenislamarre commented:
for the KEY_ BTN_. the problem is that when configuring,
1) i press on BTN_LEFT
2) dolphin saves LEFT
3) at file rereading, dolphin things it is KEY_LEFT because KEY_LEFT is before BTN_LEFT in input.h
some other buttons are concerned. KEY_0/BTN_0 and so on.
Yes, that would break some configs if LEFT is not replaced by KEY_LEFT (and so on). it is a choice between fixing or keeping.

"const bool is_pointing_device = libevdev_has_property(dev, INPUT_PROP_BUTTONPAD);" looks very suspicious to me. (while the name of the variable doesn't correspond to the property checked). it is more for touchpads.

"const bool has_sensible_button_names = !has_btn_joystick;" looks like an assumption that joystick drivers are less correctly written that other inputs. looks strange too. i don't get what this one (cosmetic too ?)

https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/InputCommon/ControllerInterface/evdev/evdev.cpp#L555
this is this return which is braking my gun. (together with the BTN_LEFT/KEY_LEFT)
because the AddAnalogInputs loop is not run.
in fact i must confess i don't understand why in dolphin, for some devices we do :
add_cursor_input(ABS_X);
while on some others it is :
AddAnalogInputs(new Axis(num_axis, axis, false, dev), new Axis(num_axis, axis, true, dev));
should we be able to simulate ir with a touchpad ?

jordan-woyak commented:
The `CursorInput` main difference is that it is not automatically detectable in the mapping dialog. It requires use of the right-click-menu advanced dialog to bind. This is done for accelerometer, gyroscope, and absolute inputs so they don't randomly bind when users are moving their mouse or physical gamepad around.

These things that you removed are definitely dirty hacks, but they weren't added there for fun. They fixed a a number of devices. You've undone that to fix one device.

Hopefully tonight I'll get around to cleaning up my local changes that I mentioned. The changes are related to input name collisions and evdev node combining. I'll take into account what you've done here and see if I can fix your gun without breaking other things.

jordan-woyak commented:
@nadenislamarre
Can you please show the output of `evtest` when your gun is attached. I'm looking for the equivalent of this: https://pastebin.com/raw/4FZDyEe8

Also, please notice the nonsensical button codes in my link. Stuff like that is the reason for `has_sensible_button_names`.

nadenislamarre commented:
[joysticks.tar.gz](https://github.com/dolphin-emu/dolphin/files/9992924/joysticks.tar.gz)

Hi,
please find the attached files with evtest when 2 of my guns are plugged.
however,
note that when you plug a usb gun or a wiimote, this is not 1 event. but generally 3 for the different parts of the gun/wiimote.
thus, to make it work with dolphin, we use evsieve to merge devices.

the aim of my pr was to share and eventually let you take some things you find interesting. but i can understand that my query is too specific. On batocera, our build system includes .patch files applied automatically on top of program, so, it is not needed to modify for us ; just consider these changes if you think it is good for the project in general.

jordan-woyak commented:
Thank you,

Which one of these event files has axes that aren't exposed because the `AddAnalogInputs` loop isn't run?

nadenislamarre commented:
ABS_X and ABS_Y

jordan-woyak commented:
> ABS_X and ABS_Y

Sorry, which device file has the issue? Dolphin should pick up just `ABS_X` and `ABS_Y` fine (and call them Cursor X/Y).

nadenislamarre commented:
1) the raw devices of the sinden gun
evtest.event11.txt:Input device name: "SindenCamA: SindenCamA"
evtest.event12.txt:Input device name: "Unknown SindenLightgun Keyboard"
evtest.event17.txt:Input device name: "Unknown SindenLightgun Mouse"

2) the merged devices of the sinden gun : the final one used in dolphin
evtest.event19.txt:Input device name: "Sinden lightgun"

3) the raw devices of the wiimote
evtest.event21.txt:Input device name: "Nintendo Wii Remote Accelerometer"
evtest.event22.txt:Input device name: "Nintendo Wii Remote IR"
evtest.event23.txt:Input device name: "Nintendo Wii Remote"

4) the merged devices of the wiimote
evtest.event24.txt:Input device name: "wiigun"

5) the final wiimote device used in dolphin, it is a virtual device created on top of the merged device of the wiimote to add calibration
evtest.event25.txt:Input device name: "wiigun calibrated"

nadenislamarre commented:
for information,
in batocera, i disabled merging of evdev devices.

It is not very compatible with autoconfiguration of emulator that we do. because the axis number are no more the same one that other emulators does.

Main Stick/Left = `Axis 0-`
Main Stick/Right = `Axis 0+`
Main Stick/Up = `Axis 1-`
Main Stick/Down = `Axis 1+`

I understand that you do events merging, but for me, it should be an option. it is probably wanted for the ps4 pad to bypass the choice made by the ps4 driver developper who choose to make it 2 devices, but for the rest, i think it causes issues depending on the pads. I think that the correct way it to allow for each button input from different devices
Main Stick/Left = `Joystick2 Axis 0-` and so on.
